src/cargo/core/compiler/unit_dependencies.rs

```diff
@@ -142,8 +144,27 @@ pub fn build_unit_dependencies<'a, 'gctx>(
     // which affect the determinism of the build itself. As a result be sure
     // that dependency lists are always sorted to ensure we've always got a
     // deterministic output.
-    for list in state.unit_dependencies.values_mut() {
-        list.sort();
+    for (unit, list) in &mut state.unit_dependencies {
+        let is_multiple_build_scripts_enabled = unit
+            .pkg
+            .manifest()
+            .unstable_features()
+            .require(Feature::multiple_build_scripts())
+            .is_ok();
+
+        // We do not sort build scripts to preserve the user-defined order.
+        // It ensures deterministic behaviour as well as user control over build scripts order.
+        if is_multiple_build_scripts_enabled {
+            list.sort_by_key(|unit_dep| {
+                if unit_dep.unit.target.is_custom_build() {
+                    (0, None)
```

Comment:
Code organization wise, a comment like this belongs inside of the `if is_custom_build()` branch as its explaining that branch.   The connotation for this location is its explaining the nightly check.  Inside of this `if` would be for explaining the overall logic.

```suggestion
        if is_multiple_build_scripts_enabled {
            list.sort_by_key(|unit_dep| {
                if unit_dep.unit.target.is_custom_build() {
                    // We do not sort build scripts to preserve the user-defined order.
                    //
                    // In terms of determinism, we are assuming nothing interferes with order from when the user set it in `Cargo.toml` to here
                    (0, None)
```
(also a small tweak to the determinism comment)